CP-311608: Add elf note to check kernel supports hypercall filtering

From: Frediano Ziglio <frediano.ziglio@citrix.com>

This is required for Shim review.

Signed-off-by: Frediano Ziglio <frediano.ziglio@citrix.com>

diff --git a/xen/arch/x86/dom0_build.c b/xen/arch/x86/dom0_build.c
index 6e30acce9e2c..964d644930ee 100644
--- a/xen/arch/x86/dom0_build.c
+++ b/xen/arch/x86/dom0_build.c
@@ -7,6 +7,7 @@
 #include <xen/init.h>
 #include <xen/iocap.h>
 #include <xen/libelf.h>
+#include <xen/lockdown.h>
 #include <xen/param.h>
 #include <xen/pfn.h>
 #include <xen/sched.h>
@@ -330,6 +331,27 @@ int __init initdom_check_parms(
         return -EINVAL;
     }
 
+    if ( is_locked_down() )
+    {
+        uint64_t filter_version =
+            parms->xs_elf_notes[XS_ELFNOTE_PRIVCMD_FILTERING].data.num;
+
+        if ( filter_version == 0 )
+        {
+            printk(XENLOG_ERR
+                   "Kernel hypercall filtering version is missing\n");
+            return -EINVAL;
+        }
+        if ( filter_version != PRIVCMD_FILTERING_ABI_VERSION )
+        {
+            printk(XENLOG_ERR
+                   "Kernel hypercall filtering version %#" PRIx64
+                   " is not compatible with expected %#lx\n",
+                   filter_version, PRIVCMD_FILTERING_ABI_VERSION);
+            return -EINVAL;
+        }
+    }
+
     return 0;
 }
 
diff --git a/xen/common/libelf/libelf-dominfo.c b/xen/common/libelf/libelf-dominfo.c
index 387ea2909ee0..4dd3a2a3ea1f 100644
--- a/xen/common/libelf/libelf-dominfo.c
+++ b/xen/common/libelf/libelf-dominfo.c
@@ -278,6 +278,7 @@ elf_errorstatus elf_xs_parse_note(struct elf_binary *elf,
         bool str;
     } note_desc[] = {
         [XS_ELFNOTE_PV_IOMMU] = { "XS_PV_IOMMU", 0},
+        [XS_ELFNOTE_PRIVCMD_FILTERING] = { "XS_PRIVCMD_FILTERING", 0},
     };
 /* *INDENT-ON* */
 
diff --git a/xen/include/public/elfnote.h b/xen/include/public/elfnote.h
index e2743f74c195..3d1b7bcea843 100644
--- a/xen/include/public/elfnote.h
+++ b/xen/include/public/elfnote.h
@@ -283,7 +283,14 @@
  */
 #define XS_ELFNOTE_PV_IOMMU               0
 
-#define XS_ELFNOTE_MAX XS_ELFNOTE_PV_IOMMU
+/*
+ * If this is present, Dom0 support hypercall filtering,
+ * needed for Secure Boot.
+ * The number contains a version of the filter.
+ */
+#define XS_ELFNOTE_PRIVCMD_FILTERING      1
+
+#define XS_ELFNOTE_MAX XS_ELFNOTE_PRIVCMD_FILTERING
 
 #endif /* __XEN_PUBLIC_ELFNOTE_H__ */
 
diff --git a/xen/include/xen/lockdown.h b/xen/include/xen/lockdown.h
index b2baa31caac8..39b51f77092a 100644
--- a/xen/include/xen/lockdown.h
+++ b/xen/include/xen/lockdown.h
@@ -6,4 +6,19 @@
 bool is_locked_down(void);
 void lockdown_init(const char *cmdline);
 
+/*
+ * Filtering code versioning.
+ * Dom0 kernel should define this in the ELF notes.
+ * Xen, if Secure Boot (or lockdown) is enabled will check that the
+ * version is in a given range; this to prevent loading kernel not
+ * properly filtering hypercalls from userspace (potentially possible
+ * in case of OEM signed kernels).
+ *
+ * The current schema is 0xMMmmnn where "MM" is the XenServer major "mm"
+ * is XenServer minor and "nn" is an incremental number.
+ * If we update the hypercall ABI in a way that is not compatible with
+ * the kernel filtering code "nn" should be increased.
+ */
+#define PRIVCMD_FILTERING_ABI_VERSION 0x090003UL
+
 #endif /* XEN__LOCKDOWN_H */
